runners/direct-java/src/main/java/org/apache/beam/runners/direct/DoFnLifecycleManager.java

```diff
@@ -58,7 +59,7 @@ private DoFnLifecycleManager(DoFn<?, ?> original) {
 
   public DoFn<?, ?> get() throws Exception {
     Thread currentThread = Thread.currentThread();
-    return outstanding.get(currentThread);
+    return outstanding.get(currentThread).getDoFn();
```

Comment:
Ah, I see. This may seem like scope creep, but the real goal is to thread this all the way through, so actually we should get rid of that call. I think you can do that by changing `ParDoEvaluator` to use `DoFn` here: https://github.com/apache/incubator-beam/blob/master/runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java#L49

That, in turn, is blocked by the use of `ParDoEvaluator` in GABW, which @tgroh knows about and thinks should go away (by just running the gabwDoFn directly), here: https://github.com/apache/incubator-beam/blob/4cb1d10dfb49c7a0622bfe8ec7b5524295ef6da1/runners/direct-java/src/main/java/org/apache/beam/runners/direct/GroupAlsoByWindowEvaluatorFactory.java#L124